compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs

```diff
@@ -534,13 +537,95 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 let ty = self.tcx.erase_late_bound_regions(ty);
                 if self.can_coerce(expected, ty) {
                     err.span_label(sp, format!("expected `{}` because of return type", expected));
+                    self.try_suggest_return_impl_trait(err, expected, ty, fn_id);
                     return true;
                 }
                 false
             }
         }
     }
 
+    /// check whether the return type is a generic type with a trait bound
+    /// only suggest this if the generic param is not present in the arguments
+    /// if this is true, hint them towards changing the return type to `impl Trait`
+    /// ```
+    /// fn cant_name_it<T: Fn() -> u32>() -> T {
+    ///     || 3
+    /// }
+    /// ```
+    fn try_suggest_return_impl_trait(
+        &self,
+        err: &mut DiagnosticBuilder<'_>,
+        expected: Ty<'tcx>,
+        found: Ty<'tcx>,
+        fn_id: hir::HirId,
+    ) {
+        debug!("try_suggest_return_impl_trait, expected = {:?}, found = {:?}", expected, found);
+
+        let ty_param = match expected.kind() {
+            ty::Param(ty_param) => ty_param,
+            _ => return,
+        };
+
+        let fn_node = self.tcx.hir().find(fn_id);
+
+        if let Some(hir::Node::Item(hir::Item {
+            kind:
+                hir::ItemKind::Fn(
+                    hir::FnSig { decl: hir::FnDecl { inputs, output, .. }, .. },
+                    hir::Generics { params, where_clause, .. },
+                    _body_id,
+                ),
+            ..
+        })) = fn_node
+        {
+            if let Some(generic_param) = params.get(ty_param.index as usize) {
+                let all_bounds = generic_param
+                    .bounds
+                    .iter()
+                    .chain(
+                        where_clause
+                            .predicates
+                            .iter()
+                            .flat_map(|p| match p {
```

Comment:
You can simplify this chain.
```suggestion
                            .filter_map(|p| match p {
```